### PR TITLE
Callback refactor

### PR DESCRIPTION
**Pull request by cosama**
First follow up to #71, with a smaller amount of changes.

This only restructures the callback, moving most of it to specific functions, as the callback is getting quite large and hard to follow the full `if else` logic.

Still quite a large diff, but a little bit better than before.

### CODE REVIEW

**Comment by cosama** [comment]
@dawonn I think this is ready. It is the smallest changes possible for the callback refactor. I will add more pull request once this one is in.

**Comment by dawonn** [comment]
Thanks for the extra effort
